### PR TITLE
Simplex example

### PR DESCRIPTION
**Pull request by duvenaud**
The goal of this demo is to show off Dex's ability to work with
structured matrices without using integers to denote slices.
This is done using records to construct index sets. 

 I started it in July 2020, and then in November 2021, @chenzizhao got it to pass all the tests.  I updated all the syntax, but there might be nicer ways to do some of the updates now that Dex has improved.

This is ready for review.

### CODE REVIEW

File: examples/simplex.dx [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+
+data ObjectiveType =
+  Maximize
+  Minimize
+
+def Constraint (n:Type) : Type =
+  {coeffs:n=>Float & offset:Float & conType:ConstraintType}
+
+def ColIx (cons:Type) (vars:Type) : Type =
+  {variables:vars | slacks:cons | value: Unit}
```
Comment (apaszke) [inline examples/simplex.dx:107]: What does cons mean here?

File: examples/simplex.dx [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+
+def eval_objective (coeffs: n=>Float) (x: n=>Float) : Float = dot coeffs x
+
+def constraints_satisfied (constraints: m=>Constraint n) (x: n=>Float) : Bool =
+  constraints_ok = all for c.
+    ({offset=offset, coeffs=coeffs, conType=conType}) = constraints.c
+    case conType of
+      LessThan    -> dot coeffs x <= offset
+      GreaterThan -> dot coeffs x >= offset
+  constraints_ok && all for i:n. x.i >= 0.0 || x.i ~~ 0.0
```
Comment (apaszke) [inline examples/simplex.dx:142]: Are you implicitly adding positivity constraints for all variables?

File: makefile [hunk 3]
```diff
@@ -115,7 +115,7 @@ dexrt-llvm: src/lib/dexrt.bc
 
 # --- running tests ---
 
-example-names = mandelbrot pi sierpinski rejection-sampler \
+example-names = simplex mandelbrot pi sierpinski rejection-sampler \
```
Comment (apaszke) [inline makefile:118]: Could you please move this example to the end? The ones at the front are usually quite simple and usually much better for debugging.

**Review by apaszke — APPROVED** [review]
Looks nice! I'd really love to see this as a literate example since it is pretty hard to follow at times if one's not intimately familiar with the simplex algorithm. The few comments help, but are not enough for me 😕 In any case, I'm good with merging this if only we move the example to the end of the list in our makefile.

**Comment by apaszke** [comment]
Also there seem to be some conflicts are make this unmergeable

File: examples/simplex.dx [hunk 4]
```diff
@@ -0,0 +1,655 @@
...
+  {coeffs=equal_a.(1@_), offset=equal_b.(1@_), conType=GreaterThan}]
+
+:p simplex dup_constraints2 equal_obj Minimize
+> (Solution [0.5, 0.])
+
+
+-- the first constraint is always satisfied
+redundant_obj = [1.0, 1.0]
+redundant_constraints =
+  [{coeffs=[0.0, 0.0], offset=0.0, conType=LessThan},
```
[the remark this answers is not in the document]
Comment (chenzizhao) [reply examples/simplex.dx:584]: That looks nice! I will keep the explicit comparators for now but thank you for the suggestion :blush:

File: examples/simplex.dx (hunk 1 above)
Comment (chenzizhao) [reply examples/simplex.dx]: That's constraints here. I added comments for clarity.

File: examples/simplex.dx (hunk 2 above)
Comment (chenzizhao) [reply examples/simplex.dx]: Yep you are right! That is one assumption of this "basic" implementation. There was a note about this in the main `simplex` function, but I guess that's a long shot for any reader. I've added a few comments for readability.

File: makefile (hunk 3 above)
Comment (chenzizhao) [reply makefile]: Done.

**Comment by chenzizhao** [comment]
Thank you for the feedback! I made a couple of revisions:
- Compare an example of the tableau and its Dex type side by side. Hopefully it reads less dry now.
Screenshot from 2021-12-23 03-32-37
- Add a summary of the simplex procedure and map each step to a subsection of the gigantic "Simplex helper functions" section. This should help establish what to expect for those who are not as familiar with the simplex method using tableaus, and make it easier to navigate too.

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Awesome, thank you!
